pkg/scheduler/service/cleaner.go

```diff
@@ -67,154 +69,194 @@ func (c *cleaner) purgeReconciliations(transition *ClusterStatusTransition, conf
 
 //Purges reconciliations using rules from: https://github.com/kyma-incubator/reconciler/issues/668
 func (c *cleaner) purgeReconciliationsNew(transition *ClusterStatusTransition, config *CleanerConfig) {
-	now := time.Now()
 
-	latestReconciliations, err := c.getLatestReconciliations(config.keepLatestEntitiesCount(), transition)
+	clusters, err := transition.inventory.GetAll()
```

Comment:
I think this can be a very expensive operation if we get all clusters, is there any way we can instead use a query here that already takes into account the age through config.maxEntitiesAgeDays() and the count and status criteria? Maybe I overestimate this, but I think that getting all clusters can have thousands of records through which you iterate both deleteRecordsByAge and deleteRecordsByCountAndStatus, even if they end up as NoOp. Maybe a delete driven by the database is more efficient here. Let me know what you think :) I see that you are using these methods maybe also for readability and configurability but I think the delete could also be written as one single query

Reply:
@jakobmoellersap: TLDR - We will improve things with extensions to the ReconciliationRepository (currently under development). The logic that is currently used is not much better, it also removes records one by one, but it doesn't have to list the clusters. Still I think after the improvements we'll run faster than current logic.

We want to use our object model (simple ORM), not direct queries, unless it's the only option.
So we are extending the functionality of ReconciliationRepository to make this code more efficient in the future - it's already reflected in the functions composition.
We could probably achieve a "single-delete statement" solution here even with the object model, but It would require substantial coding effort and it may be not necessary. Perhaps just the single-delete statement of entries older than "Y" days is enough to cut down the record set to a reasonable amount, e.g: 100 per cluster.
Then we can filter this out in the memory and remove remaining records with another single-delete statement using: 'WHERE IN ("id1", "id2", ...') syntax.

As for the clusters you are right, there may be many of them, I'd expect up to 500 per environment.
But we always have to delete records in the context of the single cluster. I don't see any other way of doing that, except of some stored procedure that runs entirely in the database (we don't want it for now AFAIK).

Finally, the cleaner is a cron job and it is expected to be a long-running process, so maybe it's not an issue. If, after improvements, we can process a cluster in a 10 seconds (easily doable), then we can process 360 clusters in an hour - that should be enough.

Reply:
I totally agree with the TLDR.

Without extending further on it, I think this "ORM" (which i think is just a row mapper since we do not manage relations) is already being challenged as low prio issue, so perfectly ok that we limit our effort here for now.

I think the stored procedure is not the only way to optimize deletes here, but this we can also talk about again then in detail when it comes to our other optimisation efforts on our DB layer in general.

Thanks for the explanation!

Reply:
@jakobmoellersap One more thing. If this approach is not efficient enough, I'd consider a view-based approach. We're currently introducing views to optimize some other cleaning process. Once it's established, we can do the same here.
But I think it won't be necessary here. We'll see.